apx/server/src/apx_server.c

```diff
@@ -110,7 +120,8 @@ static void apx_server_accept(void *arg,msocket_server_t *srv,msocket_t *msocket
          msocket_handler_t handlerTable;
 
          //add it to our list of connections. The linked list is used to keep track of all open connections
-         adt_list_insert(&self->connections,newConnection);
+         MUTEX_LOCK(self->mutex);
+         adt_list_insert(&self->connections, newConnection);

```

Comment:
I'm pretty sure you can release the lock after this line. The remaining stuff in this code block should affect only newConnection.

Reply:
The registration with the nodemanager needs protection as well.

If two connections are made with the same nodename, or a disconnect occurs before registration with the nodemanager I thought it would be more consistent if both connection tracking AND nodemanager registration was made with lock held.

Reply:
You're absolutely right. It happens inside the call to apx_serverConnection_start where there is another bunch of variables (inside nodeManager) that lacks proper thread-protection.